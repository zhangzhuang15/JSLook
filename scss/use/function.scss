@use "sass:string";

@function helloColor($msg) {
    @if string.index($msg, "hello") > -1 {
        @return "black";
    } 

    @return "white";
}