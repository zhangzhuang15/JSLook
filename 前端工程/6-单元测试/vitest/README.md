## Get Started
### run unit test
```sh 
$ pnpm unit-test
```
#### run benchmark test 
```sh 
$ pnpm bench-test
```
#### run coverage test 
```sh 
$ pnpm coverage
```

#### how to create this demo 
1. `pnpm install -D vitest`
2. write your `vitest.config.ts` following vitest official intro
3. write your source code, try to test them

there's no scaffold like `pnpm create vite`