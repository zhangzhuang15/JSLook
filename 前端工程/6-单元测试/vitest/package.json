{
  "name": "vitest",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "unit-test": "vitest",
    "bench-test": "vitest bench",
    "coverage": "vitest run --coverage",
    "typecheck-test": "vitest typecheck",
    "check": "tsc --noEmit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@vitest/coverage-v8": "^0.34.3",
    "typescript": "^5.2.2",
    "vitest": "^0.34.3"
  }
}
