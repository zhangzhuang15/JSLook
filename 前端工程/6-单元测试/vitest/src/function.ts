/// <reference path="../node_modules/vitest/importMeta.d.ts" />

// To get type hints of `import.meta.vitest`, add the special comment above.
// We don't try to modify `types` of tsconfig.json to do the same thing,
// because we will lose the auto-imported type under @types, e.g. we lose
// the type hints of `@types/node`.

// `import.meta` only has 2 attributes, url and resolve. `vitest` is
// supported by vite in compiling time. `import.meta.vitest` will be
// replaced of other valid codes, such as a valid javascript Object.
// In other words, `import.meta.vitest` is only valid in type declaration,
// not in runtime stage.


export function add(a: number, b: number): number {
    return a + b;
}

export function greater(a: number, b: number): number {
    return a > b ? 1 : (a === b ? 0 : -1)
}

// 类似于Rust方式，在源码内嵌入测试代码, Great!
if (import.meta.vitest) {
    const { test, expect } = import.meta.vitest
    
    test("test add method in source way like Rust", () => {
        expect(add(1,2)).toBe(3)
    })
}