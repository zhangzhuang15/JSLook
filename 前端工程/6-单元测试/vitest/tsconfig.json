{
    "compilerOptions": {
        "module": "ESNext",
        "moduleResolution": "NodeNext",
        "paths": {
            "@/*": ["./src/*"]
        },
    },
    "include": ["./src/**/*.ts", "./test/**/*.ts"]
}