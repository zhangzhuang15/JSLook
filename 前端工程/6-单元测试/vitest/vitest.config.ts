// ref: https://vitest.dev/config/#configuration

import { defineConfig } from "vitest/config"
import path from "path"


export default defineConfig({
    test: {
        /**
         * oh my god! vitest is too stupid!ðŸ¤¦
         * 
         * if you write { "@/*": "./src/*" }, vitest cannot find our .ts file under src directoryðŸ˜¨
         * 
         * ref: https://juejin.cn/post/7139463114088513567
         */
        alias: [
            {
                find: /@\/(.*)/,
                replacement: path.join(__dirname, "src", "$1")
            }
        ],
        include: ["./test/*.{test,spec}.ts"],
        // "./src/**/*.ts" for in source test case which
        // vitest supports like Rust
        includeSource: ["./src/**/*.ts"],

        coverage: {
            provider: "v8"
        },

        /**
         * benchmark test config
         */
        benchmark: {
            "include": ["./test/*.bench.ts"]
        },

        /**
         * typecheck test config
         */
        typecheck: {
            include: ["./test/*.type.ts"]
        }

    },

    define: {
        // there are test codes in source code, we have
        // to delete those codes in production, so do this:
        "import.meta.vitest": undefined
    }
})


