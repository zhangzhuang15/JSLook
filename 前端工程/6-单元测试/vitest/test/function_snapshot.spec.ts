import { expect, test } from "vitest"
import { add } from "@/function"

test("add 1 and 2 should be 3", () => {
    expect(add(1, 2)).toMatchSnapshot()
})

test("add 2 and 3 should be 5", () => {
    expect(add(2, 3)).toMatchInlineSnapshot('5')
})