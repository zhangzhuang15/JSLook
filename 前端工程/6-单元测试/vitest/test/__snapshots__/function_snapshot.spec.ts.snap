// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add 1 and 2 should be 3 1`] = `3`;
